vm-virtio/src/block.rs

```diff
@@ -584,7 +584,7 @@ impl<T: 'static + DiskFile + Send> VirtioDevice for Block<T> {
         self.acked_features |= v;
     }
 
-    fn read_config(&self, offset: u64, mut data: &mut [u8]) {
+    fn read_config(&mut self, offset: u64, mut data: &mut [u8]) {
```

Comment:
I'm not sure about this change. The read config is really about the driver running in the guest asking about the virtio device what is its config. This means `read_config()` should never modify the device configuration.
I understand the configuration needs to be determined through the communication with the vhost-user daemon, but that means we get this information through the vhost-user socket.

Reply:
@sboeuf , yes, but for user blk, there are some difference. The queue num and wce info are set in cloud-hypervisor cmdline, not from backend like SPDK, but those parameters are needed set into blk config, which are required in guest kernel.  For WCE, guest kernel can change this value, and set back to backend. So, read_config need mut setting here, thanks.

Reply:
Not sure I understand. I was saying this function will be exclusively triggered when the guest will read the configuration value. And in case the guest tries to change the configuration, it should go through `write_config()` instead.
BTW, what's WCE?

Reply:
WCE: Write Cache Enable, In guest kernel, "write back" and "write through" can be set to this parameter.

The user blk config info is coming from two parts, one part mostly from backend SPDK, another part from cloud-hypervisor cmdline(queue num and WCE). When we call read_config() through vhost_rs protocol to get SPDK's config info, we need add extra info into this config struct, like queue num and wce value. So, this read_config should be changed in cloud-hypervisor side.  This is difference with net or fs.

Below is user blk config info, which will be exposed to guest kernel
struct virtio_blk_config {
    capacity: u64,
    size_max: u32,
    seg_max: u32,
    cylinders: u16,
    heads: u8,
    sectors: u8,
    blk_size: u32,
    physical_block_exp: u8,
    alignment_offset: u8,
    min_io_size: u16,
    opt_io_size: u32,
    **wce: u8,**
    unused0: [u8; 1],
    **num_queues: u16,**
    max_discard_sectors: u32,
    max_discard_seg: u32,
    discard_sector_alignment: u32,
    max_write_zeroes_sectors: u32,
    max_write_zeroes_seg: u32,
    write_zeroes_may_unmap: u8,
    unused1: [u8; 3],
}

Reply:
Ok so if I follow correctly your explanation, you will get `wce` and `num_queues` information from the cloud-hypervisor command line. That's exactly what's happening for `vhost-user-net` [here](https://github.com/intel/cloud-hypervisor/blob/master/vm-virtio/src/vhost_user/net.rs#L39) and more specifically [here](https://github.com/intel/cloud-hypervisor/blob/master/vm-virtio/src/vhost_user/net.rs#L88). I know `vhost-user-net` is not directly writing those information to `config_space` byte array, but that's basically what you need to do.
Look at how it's done for virtio-pmem for a good example [here](https://github.com/intel/cloud-hypervisor/blob/master/vm-virtio/src/pmem.rs#L298-L301).

Basically, when those information come from the cloud-hypervisor command line, you can set the virtio configuration as soon as you create the instance. And if needed, you can even store those information to have them handy at any point in time.

Reply:
@sboeuf ,

The latest code remove mut from "fn read_config(& mut self, offset: u64, mut data: &mut [u8])", thanks.